dll/win32/imm32/imm.c

```diff
@@ -2950,40 +3005,134 @@ BOOL WINAPI ImmSetConversionStatus(
     return TRUE;
 }
 
+/***********************************************************************
+ *		ImmLockImeDpi (IMM32.@)
+ */
+PIMEDPI WINAPI ImmLockImeDpi(HKL hKL)
+{
+    PIMEDPI pImeDpi = NULL;
+
+    TRACE("ImmLockImeDpi(%p)\n", hKL);
+
+    RtlEnterCriticalSection(&g_csImeDpi);
+
+    if (g_pImeDpiList)
+    {
+        /* find by hKL */
+        pImeDpi = g_pImeDpiList;
+        do
+        {
+            if (pImeDpi->hKL == hKL) /* found */
+            {
+                /* lock if possible */
+                if (pImeDpi->dwFlags & IMEDPI_FLAG_UNKNOWN)
+                    pImeDpi = NULL;
+                else
+                    ++(pImeDpi->cLockObj);
+                break;
+            }
+            pImeDpi = pImeDpi->pNext;
+        } while (pImeDpi);
+    }
+
+    RtlLeaveCriticalSection(&g_csImeDpi);
+    return pImeDpi;
+}
+
+/***********************************************************************
+ *		ImmUnlockImeDpi (IMM32.@)
+ */
+VOID WINAPI ImmUnlockImeDpi(PIMEDPI pImeDpi)
+{
+    PIMEDPI pEntry, pNext;
+
+    TRACE("ImmUnlockImeDpi(%p)\n", pImeDpi);
+
+    if (pImeDpi == NULL)
+        return;
+
+    RtlEnterCriticalSection(&g_csImeDpi);
+
+    /* unlock */
+    --(pImeDpi->cLockObj);
+    if (pImeDpi->cLockObj != 0)
+    {
+        RtlLeaveCriticalSection(&g_csImeDpi);
+        return;
+    }
+
+    if ((pImeDpi->dwFlags & IMEDPI_FLAG_UNKNOWN) == 0)
+    {
+        if ((pImeDpi->dwFlags & IMEDPI_FLAG_UNKNOWN2) == 0 ||
+            (pImeDpi->dwUnknown1 & 1) == 0)
```

Comment:
Suspicious: how do you know that dwUnknown1 has to have a flag "1", if it's unknown?

Reply:
```c
        PIMEDPI pImeDpi = (PIMEDPI)calloc(sizeof(IMEDPI), 1);
        pImeDpi->cLockObj = 1;
        pv = &pImeDpi->dwUnknown1;
        printf("pv: %p\n", pv);
        fflush(stdout);
        DebugBreak();
        (*pImmUnlockImeDpi)(pImeDpi);
```
![image](https://user-images.githubusercontent.com/2107452/126065451-93235667-4903-4d4f-92cc-455769be5494.png)
![image](https://user-images.githubusercontent.com/2107452/126065456-c3e8696a-8c30-4414-a8be-71d50852b81f.png)
My assumption seems wrong. Sorry.

Reply:
UPDATE.
```c
        PIMEDPI pImeDpi = (PIMEDPI)calloc(sizeof(IMEDPI), 1);
        pImeDpi->cLockObj = 1;
        pv = &pImeDpi->dwUnknown1;
        pImeDpi->dwUnknown1 = (argc >= 2 ? atoi(argv[1]) : 0);
        printf("pv: %p\n", pv);
        fflush(stdout);
        DebugBreak();
        (*pImmUnlockImeDpi)(pImeDpi);
```
![image](https://user-images.githubusercontent.com/2107452/126067477-82770d60-b74c-4c2d-89c9-e6731666e72c.png)
![image](https://user-images.githubusercontent.com/2107452/126067482-98d280fa-0a4f-4fa8-be70-8aef952cacf7.png)
We can see the difference.

Reply:
![image](https://user-images.githubusercontent.com/2107452/126067640-e8f35c06-296f-47f1-9197-0e1f35b61f05.png)
![image](https://user-images.githubusercontent.com/2107452/126067652-46733624-d7b7-425b-b2fd-9be8cfaed066.png)
![image](https://user-images.githubusercontent.com/2107452/126067673-ae8e3cff-6e6d-4293-9863-da8c289ce904.png)

Reply:
`RtlLeaveCriticalSection`, `RtlHeapFree`, `RtlLeaveCriticalSection`, `RtlHeapFree`, ...
These alternate and continue.

Reply:
I have a superb custom debugger that is specialized for x86 reverse engineering. This is also harvest of it.